pkg/internal/webhookgenerator/server.go

```diff
@@ -0,0 +1,113 @@
+	CertDir string
+
+	// MutatingWebhookConfigName is the name that used for creating the MutatingWebhookConfiguration object.
+	MutatingWebhookConfigName string
+	// ValidatingWebhookConfigName is the name that used for creating the ValidatingWebhookConfiguration object.
+	ValidatingWebhookConfigName string
+
+	// Secret is the location for storing the certificate for the admission server.
+	// The server should have permission to create a secret in the namespace.
+	// This is optional. If unspecified, it will write to the filesystem.
+	// It the secret already exists and is different from the desired, it will be replaced.
+	Secret *apitypes.NamespacedName
+
+	// Service is k8s service fronting the webhook server pod(s).
+	// This field is optional. But one and only one of Service and Host need to be set.
+	// This maps to field .webhooks.getClientConfig.service
+	// https://github.com/kubernetes/api/blob/183f3326a9353bd6d41430fc80f96259331d029c/admissionregistration/v1beta1/types.go#L260
+	Service *Service
+	// Host is the host name of .webhooks.clientConfig.url
+	// https://github.com/kubernetes/api/blob/183f3326a9353bd6d41430fc80f96259331d029c/admissionregistration/v1beta1/types.go#L250
+	// This field is optional. But one and only one of Service and Host need to be set.
+	// If neither Service nor Host is unspecified, Host will be defaulted to "localhost".
+	Host *string
+}
+
+// Service contains information for creating a service
+type Service struct {
+	// Name of the service
+	Name string
+	// Namespace of the service
+	Namespace string
+	// Selectors is the selector of the service.
+	// This must select the pods that runs this webhook server.
+	Selectors map[string]string
+}
+
+// Webhook defines the basics that a webhook should support.
+type Webhook interface {
+	// GetName returns the name of the webhook.
+	GetName() string
+	// GetPath returns the path that the webhook registered.
+	GetPath() string
+	// GetType returns the Type of the webhook.
+	// e.g. mutating or validating
+	GetType() types.WebhookType
+	// Validate validates if the webhook itself is valid.
+	// If invalid, a non-nil error will be returned.
+	Validate() error
+}
+
+// Register validates and registers webhook(s) in the server
+func (s *ServerOptions) Register(webhooks ...Webhook) error {
+	for i, webhook := range webhooks {
+		// validate the webhook before registering it.
+		err := webhook.Validate()
+		if err != nil {
+			return err
+		}
+		_, found := s.registry[webhook.GetPath()]
+		if found {
+			return fmt.Errorf("can't register duplicate path: %v", webhook.GetPath())
+		}
+		s.registry[webhook.GetPath()] = webhooks[i]
+	}
+	return nil
+}
+
+// Start runs the server.
+// It will install the webhook related resources depend on the server configuration.
+func (s *ServerOptions) Start(stop <-chan struct{}) error {
+	s.setDefault()
+
+	return s.InstallWebhookManifests()
+}
```

Comment:
Same comment as above. Some of these methods `Register` and `Start` doesn't seem relevant from purely generation perspective.